// warp-contracts/lib/cjs/core/modules/impl/handler/JsHandlerApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsHandlerApi = void 0;
const utils_1 = require("../../../../utils/utils");
const LexicographicalInteractionsSorter_1 = require("../LexicographicalInteractionsSorter");
const AbstractContractHandler_1 = require("./AbstractContractHandler");
const INIT_FUNC_NAME = '__init';
const throwErrorWithName = (name, message) => {
    const error = new Error(message);
    error.name = name;
    throw error;
};
var KnownErrors;
(function (KnownErrors) {
    KnownErrors["ContractError"] = "ContractError";
    KnownErrors["ConstructorError"] = "ConstructorError";
})(KnownErrors || (KnownErrors = {}));
class JsHandlerApi extends AbstractContractHandler_1.AbstractContractHandler {
    constructor(swGlobal, contractDefinition, 
    // eslint-disable-next-line
    contractFunction) {
        super(swGlobal, contractDefinition);
        this.contractFunction = contractFunction;
    }
    async handle(executionContext, currentResult, interactionData) {
        const { interaction, interactionTx } = interactionData;
        this.setupSwGlobal(interactionData);
        this.enableInternalWrites(executionContext, interactionTx);
        this.assertNotConstructorCall(interaction);
        return await this.runContractFunction(executionContext, interaction, currentResult.state);
    }
    // eslint-disable-next-line
    initState(state) { }
    async maybeCallStateConstructor(initialState, executionContext) {
        var _a;
        if ((_a = this.contractDefinition.manifest) === null || _a === void 0 ? void 0 : _a.evaluationOptions.useConstructor) {
            const interaction = {
                input: { function: INIT_FUNC_NAME, args: initialState },
                caller: this.contractDefinition.owner
            };
            const interactionTx = { ...this.contractDefinition.contractTx, sortKey: LexicographicalInteractionsSorter_1.genesisSortKey };
            // this is hard corded sortKey to make KV possible
            const interactionData = { interaction, interactionTx };
            this.setupSwGlobal(interactionData);
            this.disableInternalWritesForConstructor();
            const result = await this.runContractFunction(executionContext, interaction, {});
            if (result.type !== 'ok') {
                throw new Error(`Exception while calling constructor: ${JSON.stringify(interaction)}:\n${result.errorMessage}`);
            }
            return result.state;
        }
        else {
            return initialState;
        }
    }
    assertNotConstructorCall(interaction) {
        var _a;
        if (((_a = this.contractDefinition.manifest) === null || _a === void 0 ? void 0 : _a.evaluationOptions.useConstructor) &&
            interaction.input['function'] === INIT_FUNC_NAME) {
            throw new Error(`You have enabled {useConstructor: true} option, so you can't call function ${INIT_FUNC_NAME}`);
        }
    }
    disableInternalWritesForConstructor() {
        const templateErrorMessage = (op) => `Can't ${op} foreign contract state: Internal writes feature is not available in constructor`;
        this.swGlobal.contracts.readContractState = () => throwErrorWithName('ConstructorError', templateErrorMessage('readContractState'));
        this.swGlobal.contracts.write = () => throwErrorWithName('ConstructorError', templateErrorMessage('write'));
        this.swGlobal.contracts.refreshState = () => throwErrorWithName('ConstructorError', templateErrorMessage('refreshState'));
        this.swGlobal.contracts.viewContractState = () => throwErrorWithName('ConstructorError', templateErrorMessage('viewContractState'));
    }
    async runContractFunction(executionContext, interaction, state) {
        const stateClone = (0, utils_1.deepCopy)(state);
        const { timeoutId, timeoutPromise } = (0, utils_1.timeout)(executionContext.evaluationOptions.maxInteractionEvaluationTimeSeconds);
        try {
            await this.swGlobal.kv.open();
            const handlerResult = await Promise.race([timeoutPromise, this.contractFunction(stateClone, interaction)]);
            if (handlerResult && (handlerResult.state !== undefined || handlerResult.result !== undefined)) {
                await this.swGlobal.kv.commit();
                return {
                    type: 'ok',
                    result: handlerResult.result,
                    state: handlerResult.state || stateClone
                };
            }
            // Will be caught below as unexpected exception.
            throw new Error(`Unexpected result from contract: ${JSON.stringify(handlerResult)}`);
        }
        catch (err) {
            await this.swGlobal.kv.rollback();
            switch (err.name) {
                case KnownErrors.ContractError:
                    return {
                        type: 'error',
                        errorMessage: err.message,
                        state: state,
                        result: null
                    };
                case KnownErrors.ConstructorError:
                    throw Error(`ConstructorError: ${err.message}`);
                default:
                    return {
                        type: 'exception',
                        errorMessage: `${(err && err.stack) || (err && err.message) || err}`,
                        state: state,
                        result: null
                    };
            }
        }
        finally {
            if (timeoutId) {
                clearTimeout(timeoutId);
            }
            await this.swGlobal.kv.close();
        }
    }
    setupSwGlobal({ interaction, interactionTx }) {
        this.swGlobal._activeTx = interactionTx;
        this.swGlobal.caller = interaction.caller; // either contract tx id (for internal writes) or transaction.owner
    }
    enableInternalWrites(executionContext, interactionTx) {
        this.assignReadContractState(executionContext, interactionTx);
        this.assignViewContractState(executionContext);
        this.assignWrite(executionContext);
        this.assignRefreshState(executionContext);
    }
}
exports.JsHandlerApi = JsHandlerApi;
