// warp-contracts/lib/cjs/core/modules/impl/handler/WasmHandlerApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WasmHandlerApi = void 0;
const AbstractContractHandler_1 = require("./AbstractContractHandler");
class WasmHandlerApi extends AbstractContractHandler_1.AbstractContractHandler {
    constructor(swGlobal, 
    // eslint-disable-next-line
    contractDefinition, wasmExports) {
        super(swGlobal, contractDefinition);
        this.wasmExports = wasmExports;
    }
    async handle(executionContext, currentResult, interactionData) {
        try {
            const { interaction, interactionTx } = interactionData;
            this.swGlobal._activeTx = interactionTx;
            this.swGlobal.caller = interaction.caller; // either contract tx id (for internal writes) or transaction.owner
            this.swGlobal.gasLimit = executionContext.evaluationOptions.gasLimit;
            this.swGlobal.gasUsed = 0;
            this.assignReadContractState(executionContext, interactionTx);
            this.assignViewContractState(executionContext);
            this.assignWrite(executionContext);
            await this.swGlobal.kv.open();
            const handlerResult = await this.doHandle(interaction);
            await this.swGlobal.kv.commit();
            return {
                type: 'ok',
                result: handlerResult,
                state: this.doGetCurrentState(),
                gasUsed: this.swGlobal.gasUsed
            };
        }
        catch (e) {
            await this.swGlobal.kv.rollback();
            // note: as exceptions handling in WASM is currently somewhat non-existent
            // https://www.assemblyscript.org/status.html#exceptions
            // and since we have to somehow differentiate different types of exceptions
            // - each exception message has to have a proper prefix added.
            // exceptions with prefix [RE:] ("Runtime Exceptions") should break the execution immediately
            // - eg: [RE:OOG] - [RuntimeException: OutOfGas]
            // exception with prefix [CE:] ("Contract Exceptions") should be logged, but should not break
            // the state evaluation - as they are considered as contracts' business exception (eg. validation errors)
            // - eg: [CE:ITT] - [ContractException: InvalidTokenTransfer]
            const result = {
                errorMessage: e.message,
                state: currentResult.state,
                result: null
            };
            if (e.message.startsWith('[RE:')) {
                this.logger.fatal(e);
                return {
                    ...result,
                    type: 'exception'
                };
            }
            else {
                return {
                    ...result,
                    type: 'error'
                };
            }
        }
        finally {
            await this.swGlobal.kv.close();
        }
    }
    initState(state) {
        switch (this.contractDefinition.srcWasmLang) {
            case 'rust': {
                this.wasmExports.initState(state);
                break;
            }
            default: {
                throw new Error(`Support for ${this.contractDefinition.srcWasmLang} not implemented yet.`);
            }
        }
    }
    async doHandle(action) {
        switch (this.contractDefinition.srcWasmLang) {
            case 'rust': {
                let handleResult = await this.wasmExports.handle(action.input);
                if (!handleResult) {
                    return;
                }
                if (Object.prototype.hasOwnProperty.call(handleResult, 'Ok')) {
                    return handleResult.Ok;
                }
                else {
                    this.logger.debug('Error from rust', handleResult.Err);
                    let errorKey;
                    let errorArgs = '';
                    if (typeof handleResult.Err === 'string' || handleResult.Err instanceof String) {
                        errorKey = handleResult.Err;
                    }
                    else if ('kind' in handleResult.Err) {
                        errorKey = handleResult.Err.kind;
                        errorArgs = 'data' in handleResult.Err ? ' ' + handleResult.Err.data : '';
                    }
                    else {
                        errorKey = Object.keys(handleResult.Err)[0];
                        errorArgs = ' ' + handleResult.Err[errorKey];
                    }
                    if (errorKey == 'RuntimeError') {
                        throw new Error(`[RE:RE]${errorArgs}`);
                    }
                    else {
                        throw new Error(`[CE:${errorKey}${errorArgs}]`);
                    }
                }
            }
            default: {
                throw new Error(`Support for ${this.contractDefinition.srcWasmLang} not implemented yet.`);
            }
        }
    }
    async maybeCallStateConstructor(initialState, executionContext) {
        var _a;
        if ((_a = this.contractDefinition.manifest) === null || _a === void 0 ? void 0 : _a.evaluationOptions.useConstructor) {
            throw Error('Constructor is not implemented for wasm');
        }
        return initialState;
    }
    doGetCurrentState() {
        switch (this.contractDefinition.srcWasmLang) {
            case 'rust': {
                return this.wasmExports.currentState();
            }
            default: {
                throw new Error(`Support for ${this.contractDefinition.srcWasmLang} not implemented yet.`);
            }
        }
    }
}
exports.WasmHandlerApi = WasmHandlerApi;
